tests/helper/helper_cmd_wrapper.go

```diff
@@ -30,11 +38,84 @@ func Cmd(program string, args ...string) *CmdWrapper {
 	}
 }
 
-func (cw *CmdWrapper) ShouldPass() *CmdWrapper {
+func (cw *CmdWrapper) Runner() *CmdWrapper {
 	fmt.Fprintln(GinkgoWriter, runningCmd(cw.Cmd))
-	session, err := gexec.Start(cw.Cmd, cw.writer, cw.writer)
-	Expect(err).NotTo(HaveOccurred())
-	cw.session = session
+	cw.session, cw.err = gexec.Start(cw.Cmd, cw.writer, cw.writer)
+	timeout := time.After(cw.timeout)
+	if cw.timeout > 0 {
+		select {
+		case <-cw.stopChan:
+			if cw.session != nil {
+				if runtime.GOOS == "windows" {
+					cw.session.Kill()
+				} else {
+					cw.session.Terminate()
+				}
+			}
+		case <-timeout:
+			if cw.session != nil {
+				if runtime.GOOS == "windows" {
+					cw.session.Kill()
+				} else {
+					cw.session.Terminate()
+				}
+			}
+		}
+	} else if cw.maxRetry > 0 {
+		cw.session.Wait()
+		// if exit code is 0 which means the program succeeded and hence we retry
+		if !cw.pass {
+			if cw.session.ExitCode() == 0 {
+				time.Sleep(time.Duration(cw.intervalSeconds) * time.Second)
+				cw.maxRetry = cw.maxRetry - 1
+				cw.Runner()
+			} else {
+				return cw
+			}
+		} else {
+			if cw.session.ExitCode() != 0 {
+				time.Sleep(time.Duration(cw.intervalSeconds) * time.Second)
+				cw.maxRetry = cw.maxRetry - 1
+				cw.Runner()
+			} else {
+				return cw
+			}
+		}
+	}
+	return cw
+}
+
+func (cw *CmdWrapper) WithRetry(maxRetry int, intervalSeconds time.Duration) *CmdWrapper {
+	cw.maxRetry = maxRetry
+	cw.intervalSeconds = intervalSeconds
+	return cw
+}
+
+func (cw *CmdWrapper) ShouldPass() *CmdWrapper {
+	cw.pass = true
+	cw.Runner()
+	Expect(cw.err).NotTo(HaveOccurred())
+	Eventually(cw.session).Should(gexec.Exit(0), runningCmd(cw.session.Command))
+	return cw
+}
+
+func (cw *CmdWrapper) ShouldFail() *CmdWrapper {
+	cw.pass = false
+	cw.Runner()
+	Consistently(cw.session).ShouldNot(gexec.Exit(0), runningCmd(cw.session.Command))
+	return cw
+}
+
+func (cw *CmdWrapper) WithTerminate(timeoutAfter time.Duration, stop chan bool) *CmdWrapper {
+	cw.timeout = time.Duration(timeoutAfter) * time.Second
+	cw.stopChan = stop
+	cw.Runner()
```

Comment:
I would suggest that everything with `With` shouldnt run the test, it should as you can observe WithRetry function.
I think we should have functions `ShouldPass` and `ShouldFail` do the run?
Also maybe throw in a `ShouldRun` which doesn’t check for output and just runs ( just proxy to Runner function)
This would provide a nice convention

Reply:
understood!